http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala

```diff
@@ -0,0 +1,202 @@
+
+abstract class ClusterTransitionRequestedMessage(updatedMachineConfig: MachineConfig, clusterId: Long)  extends LeoPubsubMessage
+
+final case class StopUpdateMessage(updatedMachineConfig: MachineConfig, clusterId: Long)
+  extends ClusterTransitionRequestedMessage(updatedMachineConfig: MachineConfig, clusterId: Long) {
+  val messageType = "stopUpdate"
+}
+
+case class ClusterTransitionFinishedMessage(clusterFollowupDetails: ClusterFollowupDetails) extends LeoPubsubMessage {
+  val messageType = "transitionFinished"
+}
+
+final case class PubsubException(message: String) extends Exception
+
+object LeoPubsubCodec {
+  implicit val stopUpdateMessageDecoder: Decoder[StopUpdateMessage] =
+    Decoder.forProduct2("updatedMachineConfig", "clusterId")(StopUpdateMessage.apply)
+
+  implicit val clusterFollowupDetailsDecoder: Decoder[ClusterFollowupDetails] =
+    Decoder.forProduct2("clusterId", "clusterStatus")(ClusterFollowupDetails.apply)
+
+  implicit val clusterTransitionFinishedDecoder: Decoder[ClusterTransitionFinishedMessage] =
+    Decoder.forProduct1("clusterFollowupDetails")(ClusterTransitionFinishedMessage.apply)
+
+  implicit val leoPubsubMessageDecoder: Decoder[LeoPubsubMessage] = {
+    for {
+      messageType <- Decoder[String].prepare(_.downField("messageType"))
+      value <- messageType match {
+        case "stopUpdate" => Decoder[StopUpdateMessage]
+        case "transitionFinished" => Decoder[ClusterTransitionFinishedMessage]
+        case other => throw PubsubException(s"found a message with an unknown type when decoding: ${other}")
+      }
+    } yield value
+  }
+
+  implicit val machineConfigEncoder: Encoder[MachineConfig] =
+    Encoder.forProduct7("numberOfWorkers",
+      "masterMachineType",
+      "masterDiskSize",
+      "workerMachineType",
+      "workerDiskSize",
+      "numberOfWorkerLocalSSDs",
+      "numberOfPreemptibleWorkers")(x => MachineConfig.unapply(x).get)
+
+  implicit val stopUpdateMessageEncoder: Encoder[StopUpdateMessage] =
+    Encoder.forProduct3("messageType", "updatedMachineConfig", "clusterId")(x => (x.messageType, x.updatedMachineConfig, x.clusterId))
+
+  implicit val clusterFollowupDetailsEncoder: Encoder[ClusterFollowupDetails] =
+    Encoder.forProduct2("clusterId", "clusterStatus")(x => (x.clusterId, x.clusterStatus))
+
+  implicit val clusterTransitionFinishedEncoder: Encoder[ClusterTransitionFinishedMessage] =
+    Encoder.forProduct2("messageType","clusterFollowupDetails")(x => (x.messageType, x.clusterFollowupDetails))
+
+  implicit val leoPubsubMessageEncoder: Encoder[LeoPubsubMessage] = (message: LeoPubsubMessage) => {
+    message.messageType match {
+      case "stopUpdate" => stopUpdateMessageEncoder(message.asInstanceOf[StopUpdateMessage])
+      case "transitionFinished" => clusterTransitionFinishedEncoder(message.asInstanceOf[ClusterTransitionFinishedMessage])
+      case other        => throw PubsubException(s"found a message with an unknown type when encoding: ${other}")
```

Comment:
you can do 
```
  implicit val leoPubsubMessageEncoder: Encoder[LeoPubsubMessage] = Encoder.instance {
    x =>
      x match {
        case  m: StopUpdateMessage => m.asJson
        case  m: ClusterTransitionFinishedMessage => m.asJson
          // add more if you have more cases
      }
  }
```

`asInstanceOf` and `throw` exception should be avoided in general